qiskit/circuit/classical/expr/constructors.py

```diff
@@ -0,0 +1,506 @@
+        type = types.Bool()
+    elif left.type.kind is types.Uint and right.type.kind is types.Uint:
+        if left.type != right.type:
+            raise TypeError(
+                "binary bitwise operations are defined between unsigned integers of the same width,"
+                f" but got {left.type.width} and {right.type.width}."
+            )
+        type = left.type
+    else:
+        raise TypeError(f"invalid types for '{op}': '{left.type}' and '{right.type}'")
+    return Binary(op, left, right, type)
+
+
+def bit_and(left: typing.Any, right: typing.Any, /) -> Expr:
+    """Create a bitwise 'and' expression node from the given value, resolving any implicit casts and
+    lifting the values into :class:`Value` nodes if required.
+
+    Examples:
+        Bitwise 'and' of a classical register and an integer literal::
+
+            >>> from qiskit.circuit import ClassicalRegister
+            >>> from qiskit.circuit.classical import expr
+            >>> expr.bit_and(ClassicalRegister(3, "c"), 0b111)
+            Binary(\
+Binary.Op.BIT_AND, \
+Var(ClassicalRegister(3, 'c'), Uint(3)), \
+Value(7, Uint(3)), \
+Uint(3))
+        """
+    return _binary_bitwise(Binary.Op.BIT_AND, left, right)
+
+
+def bit_or(left: typing.Any, right: typing.Any, /) -> Expr:
+    """Create a bitwise 'or' expression node from the given value, resolving any implicit casts and
+    lifting the values into :class:`Value` nodes if required.
+
+    Examples:
+        Bitwise 'or' of a classical register and an integer literal::
+
+            >>> from qiskit.circuit import ClassicalRegister
+            >>> from qiskit.circuit.classical import expr
+            >>> expr.bit_or(ClassicalRegister(3, "c"), 0b101)
+            Binary(\
+Binary.Op.BIT_OR, \
+Var(ClassicalRegister(3, 'c'), Uint(3)), \
+Value(5, Uint(3)), \
+Uint(3))
+    """
+    return _binary_bitwise(Binary.Op.BIT_OR, left, right)
+
+
+def bit_xor(left: typing.Any, right: typing.Any, /) -> Expr:
+    """Create a bitwise 'exclusive or' expression node from the given value, resolving any implicit
+    casts and lifting the values into :class:`Value` nodes if required.
+
+    Examples:
+        Bitwise 'exclusive or' of a classical register and an integer literal::
+
+            >>> from qiskit.circuit import ClassicalRegister
+            >>> from qiskit.circuit.classical import expr
+            >>> expr.bit_xor(ClassicalRegister(3, "c"), 0b101)
+            Binary(\
+Binary.Op.BIT_XOR, \
+Var(ClassicalRegister(3, 'c'), Uint(3)), \
+Value(5, Uint(3)), \
+Uint(3))
+    """
+    return _binary_bitwise(Binary.Op.BIT_XOR, left, right)
+
+
+def _binary_logical(op: Binary.Op, left: typing.Any, right: typing.Any) -> Expr:
+    bool_ = types.Bool()
+    left = _coerce_lossless(lift(left), bool_)
+    right = _coerce_lossless(lift(right), bool_)
+    return Binary(op, left, right, bool_)
+
+
+def logic_and(left: typing.Any, right: typing.Any, /) -> Expr:
```

Comment:
> I was thinking in the future that the short-circuit evaluation nature of the "logical" - there's no observable impact right now, but in the future with extern calls there may well be.

Hmm, yeah I'm curious how this would materialize. Is it typical to have a separate operator node for every combination of operand types? At least for the `bit_` ops, there's a sort of guarantee (assuming you use the constructor interface) that the visitor won't get called with `Uint`s of different sizes, right? But technically, the semantics of an operation on a `UInt` of some size are different from the semantics of a `UInt` of a different size. So visitor implementors would be checking the size anyway, which feels similar to requiring them to check the type family too (and possibly choosing different semantics). Which makes it seem reasonable to just have `and`, `or`, `xor` etc. expressions instead of `bit_*` and `logic_*`. I guess maybe some of this is due to the fact that we're lifting the types relatively early because we don't have precedence between the left and right sides of binary ops.

If you think the semantics of boolean operations are different enough that it's helpful to have separate expression nodes for them, then I'm fine with keeping things as is 😄.

Reply:
I can understand that point.  I do still think there's worth, even in this stage in the distinction between a Uint type and a Bool type at this level of abstraction.  `Bool` is a two-valued non-numeric type, whereas `Uint(n)` represents the infinite series of non-negative integers truncated on the additive ring to  `[0, 2 ** n)`.  It so happens that `Uint(1)` has the same number of values as `Bool`, but equally so does `Option<()>` or `Result<(), ()>` in Rust, yet they're all semantically different.

I think perhaps a better way to illustrate my unwillingness to merge the types is not on the _allowed_ operations, but on the _disallowed_ operations for Booleans.  Imo, `Bool` should not be allowed to be used in a non-equality relation (`<`, `<=`, etc), but `Uint(n)` all should be.  Looking to the future, `Bool + Bool` should be an error (or do we go to the extreme of removing duplication and disallowing one of bitwise xor and addition for `Uint(1)`?), but `Uint(1) + Uint(1)` is well defined addition on the ring.

Reply:
I should say as well: I'm super on board with this thinking about the ease of consuming the expression tree.

I personally thought that the logic that handles the `Uint(n) * Uint(n) -> Uint(n)` family of binary operations would almost certainly be parametrised in any real visitor over the generic `n`, so I wasn't really considering "look at the width of the `Uint` types" to be much by the way of operation disambiguation. I wasn't planning on defining any operations for `Uint` types that can't be described with the `n` parameter constant-generic, so in that sense, treating `Uint(1)` specially in some places (e.g. having `IfElseOp.condition` accept a `Uint(1)` expression but not a `Uint(2)`) would be a violation of that principle to me.

I also don't like the implication too much that the future subscripting operation `Uint(n) * <some int type>` would need to return `Uint(1)`.  The memory-model semantics of QC hardware ends up with single bits being the smallest addressable type for stores, so I like having a unique type that can represent that.